core/storage/email/gae_models_test.py

```diff
@@ -128,6 +149,168 @@ def test_get_by_hash_behavior_with_sent_datetime_lower_bound(self):
+
+    def setUp(self):
+        super(ReplyToIdModelUnitTests, self).setUp()
+
+        def _generate_unique_id_for_tests(unused_cls):
+            return 'Non_unique_id'
+
+        self.generate_constant_id_ctx = self.swap(
+            email_models.GeneralFeedbackEmailReplyToIdModel,
+            '_generate_unique_reply_to_id',
+            types.MethodType(
+                _generate_unique_id_for_tests,
+                email_models.GeneralFeedbackEmailReplyToIdModel))
+
+    def test_id_generation_is_correctly_formatted(self):
+        model = email_models.GeneralFeedbackEmailReplyToIdModel
+
+        created = model.create('user_id', 'thread_id')
+        self.assertEqual(created.id, 'user_id.thread_id')
+
+        created = model.create('other_user_id', 'other_thread_id')
+        self.assertNotEqual(created.id, 'user_id.thread_id')
+
+    def test_generated_reply_to_id_is_unique(self):
+        model = email_models.GeneralFeedbackEmailReplyToIdModel
+
+        all_ids = [
+            model.create('user_id', 'thread_id').reply_to_id
+            for _ in xrange(100)]
+        set_ids = set(all_ids) # Set contains unique values
+        self.assertEqual(len(all_ids), len(set_ids))
+
+    def test_create_raises_when_duplicate_ids(self):
+        with self.generate_constant_id_ctx, self.assertRaises(Exception):
+            created1 = email_models.GeneralFeedbackEmailReplyToIdModel.create(
+                'user_id', 'thread_id')
+            created1.put()
+            created2 = email_models.GeneralFeedbackEmailReplyToIdModel.create(
+                'user_id', 'thread_id')
+            created2.put()
+
+    def test_get_reply_to_id_can_be_used_to_retrieve_model(self):
+        model = email_models.GeneralFeedbackEmailReplyToIdModel
+
+        created = model.create('user_id', 'thread_id')
+        created.put()
+
+        result = model.get_by_reply_to_id(created.reply_to_id)
+
+        self.assertNotEqual(result, None) # There should be a result
+        self.assertEqual(result.reply_to_id, created.reply_to_id)
+
+    def test_nonexistent_get_reply_to_id_should_provide_no_model(self):
+        model = email_models.GeneralFeedbackEmailReplyToIdModel
+
+        created = model.create('user_id', 'thread_id')
+        created.put()
+
+        result = model.get_by_reply_to_id(created.reply_to_id)
+
+        result = model.get_by_reply_to_id('non_existent_reply_to_id')
+
+        self.assertEqual(result, None)
+
+    def test_get_can_be_used_to_retrive_model(self):
+        created = email_models.GeneralFeedbackEmailReplyToIdModel.create(
+            'user_id', 'thread_id')
+        created.put()
+
+        result = email_models.GeneralFeedbackEmailReplyToIdModel.get(
+            'user_id', 'thread_id', strict=False)
+
+        self.assertNotEqual(result, None)
+        self.assertIn('user_id.thread_id', str(result.key))
+
+    def test_get_returns_nothing_with_bad_id_if_not_strict(self):
```

Comment:
Here & elsewhere: is 'bad_id' here referring to user ID, thread ID, or both? It might be worth being specific here, and maybe having multiple tests for each of the different scenarios (bad user ID vs. bad thread ID).

Reply:
Changed names and added a few cases.